File: GFG POTD/023-October-2025/019-k-closest-values.cpp
// https://www.geeksforgeeks.org/problems/k-closest-values/1/




#include <bits/stdc++.h>
using namespace std;

class Node {
  public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};

class Solution {
  private:
    using P = pair<int, int>;
    template <typename PQ>
    
    void solve(Node *root, int tar, PQ &pq) {
        if (!root) return;
        
        solve(root -> left, tar, pq);
        
        pq.push({abs(root -> data - tar), root -> data});
        
        solve(root -> right, tar, pq);
    }
  public:
    // TC : O(N*log(N))
    // SC : O(N)
    vector<int> getKClosest(Node* root, int target, int k) {
        auto cmp = [&](P a, P b) {
            if (a.first == b.first) {
                return b.second < a.second;
            }  
            return a.first > b.first;
        };
        priority_queue<P, vector<P>, decltype(cmp)>pq(cmp);
        
        solve(root, target, pq);
        
        vector<int> result;
        while (!pq.empty() && k --) {
            result.push_back(pq.top().second);
            pq.pop();
        }
        return result;
    }
};

File: GFG POTD/023-October-2025/019-k-closest-values.java
// https://www.geeksforgeeks.org/problems/k-closest-values/1/




import java.util.*;

class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}

class Solution {
    // TC : O(N*log(N))
    // SC : O(N)
    public ArrayList<Integer> getKClosest(Node root, int target, int k) {
        Comparator<Pair> cmp = (a, b) -> {
            if (a.diff == b.diff) return Integer.compare(a.val, b.val);
            return Integer.compare(a.diff, b.diff);
        };
        PriorityQueue<Pair> pq = new PriorityQueue<>(cmp);
        
        solve(root, target, pq);
        
        ArrayList<Integer> result = new ArrayList<>();
        while (!pq.isEmpty() && k -- > 0) {
            result.add(pq.poll().val);
        }
        return result;
    }
    
    private static void solve(Node root, int tar, PriorityQueue<Pair> pq) {
        if (root == null) return;
        
        solve(root.left, tar, pq);
        
        pq.offer(new Pair(Math.abs(root.data - tar), root.data));
        
        solve(root.right, tar, pq);
    }
    
    private static class Pair {
        int diff, val;
        Pair(int diff, int val) {
            this.diff = diff;
            this.val  = val;
        }
    }
}